homeassistant/components/smarttub/controller.py

```diff
@@ -0,0 +1,137 @@
+"""Interface to the SmartTub API."""
+
+from datetime import timedelta
+import logging
+
+import async_timeout
+from smarttub import APIError, LoginFailed, SmartTub
+
+from homeassistant.config_entries import SOURCE_REAUTH
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD, CONF_SCAN_INTERVAL
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DEFAULT_SCAN_INTERVAL, DOMAIN, POLLING_TIMEOUT
+from .helpers import get_spa_name
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SmartTubController:
+    """Interface between Home Assistant and the SmartTub API."""
+
+    def __init__(self, hass):
+        """Initialize an interface to SmartTub."""
+        self._hass = hass
+        self._api = SmartTub(async_get_clientsession(hass))
+        self._account = None
+        self.spas = set()
+        self._spa_devices = {}
+
+        self.coordinator = None
+
+    async def async_setup_entry(self, entry):
+        """Perform initial setup.
+
+        Authenticate, query static state, set up polling, and otherwise make
+        ready for normal operations .
+        """
+
+        try:
+            await self._api.login(entry.data[CONF_EMAIL], entry.data[CONF_PASSWORD])
+        except LoginFailed:
+            # credentials were changed or invalidated, we need new ones
+            self._hass.async_create_task(
+                self._hass.config_entries.flow.async_init(
+                    DOMAIN,
+                    context={"source": SOURCE_REAUTH},
+                    data=None,
+                )
+            )
+
+            return False
+
+        self._account = await self._api.get_account()
+
+        self.spas = await self._account.get_spas()
+
+        self.coordinator = DataUpdateCoordinator(
+            self._hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_method=self.async_update_data,
+            update_interval=timedelta(
+                seconds=entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+            ),
+        )
+
+        await self.coordinator.async_refresh()
+
+        await self.async_register_devices(entry)
+
+        self._unsub_config_entry_listener = entry.add_update_listener(
```

Comment:
This should be in `__init__.py` and the unsubscribe callback stored in `hass.data[DOMAIN][entry.unique_id]`

Reply:
Should I just get rid of the controller class entirely and store everything in the `hass.data` hash? What's best practice here?

Reply:
There's real functionality here, I don't think you should drop it. You do need to move the update listener though.

Reply:
OK, moved it